tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,371 @@
+        {
+            var currentState = AsyncContext.Value;
+            var newState = new AsyncMethodDebuggerState
+            {
+                Parent = currentState,
+                KickoffInvocationTarget = kickoffInfo.KickoffParentObject,
+                KickoffInvocationTargetType = kickoffInfo.KickoffParentType,
+                KickoffMethod = kickoffInfo.KickoffMethod,
+                ProbeId = probeId,
+                StartTime = DateTimeOffset.UtcNow,
+                MethodMetadataIndex = methodMetadataIndex,
+                MoveNextInvocationTarget = instance
+            };
+            AsyncContext.Value = newState;
+            return newState;
+        }
+
+        /// <summary>
+        /// Restore the async context to the parent
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void RestoreContext()
+        {
+            AsyncContext.Value = AsyncContext.Value?.Parent;
+        }
+
+        /// <summary>
+        /// Begin Method Invoker
+        /// </summary>
+        /// <typeparam name="TTarget">Target object of the method. Note that it could be typeof(object) and not a concrete type</typeparam>
+        /// <param name="probeId">The id of the probe</param>
+        /// <param name="instance">Instance value</param>
+        /// <param name="methodHandle">The handle of the executing method</param>
+        /// <param name="typeHandle">The handle of the type</param>
+        /// <param name="methodMetadataIndex">The index used to lookup for the <see cref="MethodMetadataInfo"/> associated with the executing method</param>
+        /// <param name="isReEntryToMoveNext">If it the first entry to the state machine MoveNext method</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static AsyncMethodDebuggerState BeginMethod<TTarget>(string probeId, TTarget instance, RuntimeMethodHandle methodHandle, RuntimeTypeHandle typeHandle, int methodMetadataIndex, ref bool isReEntryToMoveNext)
+        {
+            if (!ProbeRateLimiter.Instance.Sample(probeId) && !isReEntryToMoveNext)
+            {
+                return AsyncMethodDebuggerState.CreateInvalidatedDebuggerState();
+            }
+
+            if (isReEntryToMoveNext)
+            {
+                // we are in a continuation, return the current state
+                return AsyncContext.Value;
+            }
+
+            isReEntryToMoveNext = true;
+            var kickoffInfo = AsyncHelper.GetAsyncKickoffMethodInfo(instance);
+            if (kickoffInfo.KickoffParentObject == null && kickoffInfo.KickoffMethod.IsStatic == false)
+            {
+                Log.Error($"{nameof(BeginMethod)}: hoisted 'this' has not found. {kickoffInfo.KickoffParentType.Name}.{kickoffInfo.KickoffMethod.Name}");
```

Comment:
Logging *anything* in this context is very dangerous as we are on the user-code critical path; if the method we are instrumenting gets executed a million times / sec, we may end up attempting to write million of logs and causing the user's request to time out (and maybe eventually run out of disk space). This is not a showstopper at the moment because we're only sampling once per second, but since both metric probes and conditionals are going to be evaluated much much more frequently, we should either address this now or come up with a plan of how to address is.

At the very least, for the time being, we should probably set `AsyncMethodDebuggerState.IsError = true` (and consider renaming that property to something that makes it clearer it denotes an error in our instrumentation rather than an error in user code).